File: book/fridgets/f#/Fridgets/FrButton.fs
namespace Fridgets

open System.Windows
open System.Windows.Input
open System.Windows.Media
open Sodium.Frp

type FrButton =
    private {
        reify : Cell<Size option> -> Stream<MouseEvent> -> Stream<KeyEvent> -> Cell<int64> -> Supply.T -> Output
        sClicked : Stream<unit>
    }
    interface IFridget with
        member this.Reify size sMouse sKey focus idSupply = this.reify size sMouse sKey focus idSupply

[<CompilationRepresentationAttribute(CompilationRepresentationFlags.ModuleSuffix)>]
module FrButton =
    let create label =
        let sClicked = sinkC <| neverS ()
        let reify size sMouse _ _ _ =

            let getPressed (e : MouseEvent) size =
                match size with
                    | None -> None
                    | Some (size : Size) ->
                        match e.args with
                            | :? MouseButtonEventArgs as b ->
                                let p = e.getPosition ()
                                if b.ChangedButton = MouseButton.Left && b.ButtonState = MouseButtonState.Pressed && p.X >= 2.0 && p.X < size.Width - 2.0 && p.Y >= 2.0 && p.Y < size.Height - 2.0
                                then Some ()
                                else None
                            | _ -> None
            let sPressed = sMouse |> snapshotC size getPressed |> filterOptionS

            let getReleased (e : MouseEvent) size =
                match size with
                    | None -> None
                    | Some (_ : Size) ->
                        match e.args with
                            | :? MouseButtonEventArgs as b ->
                                if b.ChangedButton = MouseButton.Left && b.ButtonState = MouseButtonState.Released
                                then Some ()
                                else None
                            | _ -> None
            let sReleased = sMouse |> snapshotC size getReleased |> filterOptionS

            let pressed = (sPressed |> mapToS true, sReleased |> mapToS false) |> orElseS |> holdS false
            sClicked |> sendC (sReleased |> gateC pressed)

            let typeface = Typeface(FontFamily("Helvetica"), FontStyles.Normal, FontWeights.Normal, FontStretches.Normal)
            let getDesiredSize label =
                let labelSize = FontUtilities.measureString label typeface 13.0
                Size(labelSize.Width + 14.0, labelSize.Height + 10.0)
            let desiredSize = label |> mapC getDesiredSize

            let getDrawable label size pressed =
                DrawableDelegate (fun d ->
                    match size with
                        | None -> ()
                        | Some (size : Size) ->
                            d.DrawRectangle((if pressed then Brushes.DarkGray else Brushes.LightGray), Pen(Brushes.Black, 1.0), Rect(Point(2.0, 2.0), Size(size.Width - 5.0, size.Height - 5.0)))
                            let t = FontUtilities.getStandardFormattedText label typeface 13.0 Brushes.Black
                            d.DrawText(t, Point((size.Width - t.Width) / 2.0, (size.Height - t.Height) / 2.0)))

            {
                drawable = (label, size, pressed) |> lift3C getDrawable
                desiredSize = desiredSize
                sChangeFocus = neverS ()
            }
        
        {
            reify = reify
            sClicked = sClicked |> switchS
        }

    let sClicked b = b.sClicked

File: book/fridgets/c#/Fridgets/FrButton.cs
using System.Windows;
using System.Windows.Input;
using System.Windows.Media;
using Sodium.Frp;
using Sodium.Functional;

namespace Fridgets
{
    public class FrButton : Fridget
    {
        public FrButton(Cell<string> label)
            : this(label, new StreamLoop<Unit>())
        {
        }

        private FrButton(Cell<string> label, StreamLoop<Unit> sClicked)
            : base((size, sMouse, sKey, focus, idSupply) =>
            {
                Stream<Unit> sPressed = sMouse.Snapshot(size,
                    (e, mSize) => mSize.Bind(
                        s =>
                        {
                            MouseButtonEventArgs b = e.Args as MouseButtonEventArgs;
                            Point p = e.GetPosition();
                            return b != null && b.ChangedButton == MouseButton.Left && b.ButtonState == MouseButtonState.Pressed
                                   && p.X >= 2 && p.X < s.Width - 2 && p.Y >= 2 && p.Y < s.Height - 2
                                ? Maybe.Some(Unit.Value)
                                : Maybe.None;
                        })).FilterMaybe();
                Stream<Unit> sReleased = sMouse.Snapshot(size,
                    (e, mSize) => mSize.Bind(
                        s => e.Args is MouseButtonEventArgs b && b.ChangedButton == MouseButton.Left && b.ButtonState == MouseButtonState.Released
                            ? Maybe.Some(Unit.Value)
                            : Maybe.None)).FilterMaybe();
                Cell<bool> pressed = sPressed.MapTo(true).OrElse(sReleased.MapTo(false)).Hold(false);
                sClicked.Loop(sReleased.Gate(pressed));
                Typeface typeface = new Typeface(new FontFamily("Helvetica"), FontStyles.Normal, FontWeights.Normal, FontStretches.Normal);
                Cell<Size> desiredSize = label.Map(l =>
                {
                    Size labelSize = FontUtilities.MeasureString(l, typeface, 13);
                    return new Size(labelSize.Width + 14, labelSize.Height + 10);
                });
                return new Output(
                    label.Lift(
                        size, pressed,
                        (l, mSize, p) => new DrawableDelegate(d =>
                        {
                            mSize.MatchSome(sz =>
                            {
                                d.DrawRectangle(p ? Brushes.DarkGray : Brushes.LightGray, new Pen(Brushes.Black, 1), new Rect(new Point(2, 2), new Size(sz.Width - 5, sz.Height - 5)));
                                FormattedText t = FontUtilities.GetStandardFormattedText(l, typeface, 13, Brushes.Black);
                                d.DrawText(t, new Point((sz.Width - t.Width) / 2, (sz.Height - t.Height) / 2));
                            });
                        })),
                    desiredSize,
                    Stream.Never<long>());
            })
        {
            this.SClicked = sClicked;
        }

        public Stream<Unit> SClicked { get; }
    }
}